### PR TITLE
Fixes the issue of url creation for same component when no name is provided

### LINKED ISSUE #1040: Multiple URL creation fails for same component
[kind/bug]

## What versions of software are you using?
- Operating System: Linux :heart: 
- Output of `odo version`:

`odo v0.0.16 (HEAD)`

## How did you run odo exactly?
```
$ odo create httpd
Component 'httpd-ex-httpd-ohbd' was created and ports 8080/TCP,8443/TCP were opened
To push source code to the component run 'odo push'

Component 'httpd-ex-httpd-ohbd' is now set as active component.

$ odo url create
Adding URL to component: httpd-ex-httpd-ohbd
'port' is required as the component httpd-ex-httpd-ohbd exposes 2 ports: 8080,8443

$ odo url create --port 8080
Adding URL to component: httpd-ex-httpd-ohbd
URL created for component: httpd-ex-httpd-ohbd

httpd-ex-httpd-ohbd - http://httpd-ex-httpd-ohbd-httpd-ex-pqgc-myproject.192.168.42.65.nip.io


```

## Actual behavior

```
$ odo url create --port 8443
The url httpd-ex-httpd-ohbd already exists in the application: httpd-ex-pqgc

$ odo url list
Found the following URLs for component httpd-ex-httpd-ohbd in application httpd-ex-pqgc:
NAME                    URL                                                                         PORT
httpd-ex-httpd-ohbd     http://httpd-ex-httpd-ohbd-httpd-ex-pqgc-myproject.192.168.42.65.nip.io     8080
```

## Expected behavior

```
$ odo url create --port 8443
Adding URL to component: httpd-ex-httpd-abcd
URL created for component: httpd-ex-httpd-abcd

httpd-ex-httpd-abcd - http://httpd-ex-httpd-abcd-httpd-ex-pqgc-myproject.192.168.42.65.nip.io

$ odo url list
Found the following URLs for component httpd-ex-httpd-ohbd in application httpd-ex-pqgc:
NAME                    URL                                                                         PORT
httpd-ex-httpd-ohbd     http://httpd-ex-httpd-ohbd-httpd-ex-pqgc-myproject.192.168.42.65.nip.io     8080
httpd-ex-httpd-abcd     http://httpd-ex-httpd-ohbd-httpd-ex-pqgc-myproject.192.168.42.65.nip.io     8443
```

### PR DESCRIPTION
**Pull request by mik-dass**
fixes #1040 

It introduces a new function GetValidPortNumber() to check and return the valid port of the component.

### CODE REVIEW

File: pkg/url/url.go [hunk 1]
```diff
@@ -174,3 +147,44 @@ func GetComponentServicePortNumbers(client *occlient.Client, componentName strin
...
+func GetURLName(componentName string, componentPort int) string {
+	if componentPort == -1 {
+		return componentName
+	}
+	return fmt.Sprintf("%v-%v", componentName, componentPort)
+}
+
+// GetValidPortNumber checks if the given port number is a valid component port or not
+// if port number is not provided and the component is a single port component, the component port is returned
+func GetValidPortNumber(client *occlient.Client, portNumber int, componentName string, applicationName string) (int, error) {
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: The other functions need to be modified too then, which I will do in a different PR

File: pkg/url/url.go [hunk 2]
```diff
@@ -174,3 +147,44 @@ func GetComponentServicePortNumbers(client *occlient.Client, componentName strin
...
+	componentPorts, err := GetComponentServicePortNumbers(client, componentName, applicationName)
+	if err != nil {
+		return portNumber, errors.Wrapf(err, "unable to get exposed ports for component %s", componentName)
+	}
+
+	var portFound bool
+
+	if portNumber == -1 {
+		if len(componentPorts) > 1 {
+			return portNumber, errors.Errorf("'port' is required as the component %s exposes %d ports: %s", componentName, len(componentPorts), strings.Trim(strings.Replace(fmt.Sprint(componentPorts), " ", ",", -1), "[]"))
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: I didn't understand it

File: pkg/url/url.go [hunk 3]
```diff
@@ -174,3 +147,44 @@ func GetComponentServicePortNumbers(client *occlient.Client, componentName strin
 
 	return ports, nil
 }
+
+// GetURLName returns a url name from the component name and the given port number
+func GetURLName(componentName string, componentPort int) string {
+	if componentPort == -1 {
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: done :)

File: pkg/url/url.go [hunk 4]
```diff
@@ -174,3 +147,44 @@ func GetComponentServicePortNumbers(client *occlient.Client, componentName strin
...
+		return componentName
+	}
+	return fmt.Sprintf("%v-%v", componentName, componentPort)
+}
+
+// GetValidPortNumber checks if the given port number is a valid component port or not
+// if port number is not provided and the component is a single port component, the component port is returned
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: Done :)

File: pkg/url/url.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply pkg/url/url.go]: Yes, this should be done in a different PR if we decide to do it, still not sure it's the best way to go…

File: pkg/url/url.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply pkg/url/url.go]: The error message mentions `port` which is the flag that the command uses when this function is called to pass the port number. This means that the error message in this function is tied to the calling command implementation. This should be fixed.

[the remark this answers is not in the document]
Comment (metacosm) [reply pkg/url/url.go]: More precisely, the error message should mention the missing function parameter `portNumber` and not something called `port`, which doesn't mean anything in this context.

File: pkg/url/url.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: Ok I will fix those functions in a different PR and refactor the code a bit

File: pkg/url/url.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/url/url.go]: Fixed

**Comment by mik-dass** [comment]
@metacosm Fixed Please have a look again :)

**Pull request merged**